torchtune/modules/model_fusion/_fusion.py

```diff
@@ -463,18 +500,294 @@ def forward(
+            encoder_embed = {
+                encoder: self.encoders[encoder](**encoder_input[encoder])
+                for encoder in encoder_input
+            }
+
+        # Currently, only a single encoder is supported, so we need
+        # to get the encoder key manually. When multiple encoders are
+        # supported, this can be removed.
+        decoder_encoder_input = (
+            list(encoder_embed.values())[0] if encoder_embed is not None else None
+        )
 
         output = self.decoder(
             tokens=tokens,
             mask=mask,
-            encoder_input=encoder_embed,
+            encoder_input=decoder_encoder_input,
             encoder_mask=encoder_mask,
             input_pos=input_pos,
         )
         return output
+
+
+class EarlyFusionModel(nn.Module):
+    """EarlyFusion is a type of fused model architecture where pretrained encoder(s) are combined
+    with a pretrained decoder (LLM) at the model input and not in internal layers. This is a popular architecture
+    for multimodal models, with a full overview available in `The Evolution of Multimodal Model Architectures
+    <https://arxiv.org/abs/2405.17927>`_. This module works both for decoders in which the encoder tokens are
+    inside the vocab and outside the vocab.
+
+    This module has the same methods and forward signature as :class:`~torchtune.modules.TransformerDecoder` and can be used
+    interchangeably where :class:`~torchtune.modules.TransformerDecoder` is. It combines the encoders with the decoder as a
+    single module for checkpointing and finetuning. It is expected that the encoders and decoder
+    are already defined with any extra learnable ``fusion_params``: learnable parameters to help
+    adapt the pre-trained encoders to the pre-trained decoder.
+
+    You can pass in multiple encoders as a dictionary into ``encoders``.
+
+    Note: Once the decoder is wrapped in this module, the decoder's ``tok_embeddings`` module is moved
+    to the parent EarlyFusionModel's ``tok_embeddings``. You should not forward pass the decoder individually.
+    Instead, use EarlyFusionModel's forward pass with ``encoder_input=None`` to get decoder-only outputs.
+    State dicts will automatically be updated on save and load to account for this change.
+
+    Example:
+        >>> # decoder is a text-only TransformerDecoder (e.g. llama3_8b) with no modifications
+        >>> decoder = llama3_8b()
+        >>>
+        >>> # encoder is pre-trained encoder (e.g. clip_vit_224) with an added projection head
+        >>> projection_head = FeedForward(...)
+        >>> register_fusion_module(projection_head))
+        >>> encoders = {"image": nn.Sequential(clip_vit_224(), projection_head)}
+        >>>
+        >>> # EarlyFusionModel combines the encoder and decoder
+        >>> model = EarlyFusionModel(decoder, encoders, encoder_tokens={"image": 128256})
+        >>>
+        >>> # Load full fused checkpoints
+        >>> model.load_state_dict(...)
```

Comment:
if the checkpoint is a single file you can load the entire model with encoders all at once. but I'm not sure if this will be the case for a model with multiple encoders or what the checkpoint UX would look like. I'm ok to remove this until we know for sure.